Test/dafny0/AndOrAmbiguity.dfy

```diff
@@ -6,56 +6,56 @@
 // The tests below check that these ambiguities result in understandable error message.
 
 method M0() returns (A: bool, B: bool)
-  ensures A && B ||  // error: and-or ambiguity
+  ensures A && B || A // error: and-or ambiguity
 
 method M1() returns (A: bool, B: bool)
-  ensures A || B &&  // error: and-or ambiguity
+  ensures A || B && A // error: and-or ambiguity
 
 method M2() returns (A: bool, B: bool)
-  ensures && B ||  // error: and-or ambiguity
+  ensures && B || A // error: and-or ambiguity
 
 method M3() returns (A: bool, B: bool)
-  ensures || B &&  // error: and-or ambiguity
+  ensures || B && A // error: and-or ambiguity
 
 method M4() returns (A: bool, B: bool)
-  ensures && A && B ||  // error: and-or ambiguity
+  ensures && A && B || A // error: and-or ambiguity
 
 method M5() returns (A: bool, B: bool)
-  ensures || A || B &&  // error: and-or ambiguity
+  ensures || A || B && A // error: and-or ambiguity
 
 // ---
 
 method N0() returns (A: bool, B: bool, C: bool)
```

Comment:
Nitpicking: File name only indicates ambiguities involving "and", "or", but here you also test "implication".

Reply:
Actually I think 'And' and 'Or' more often refer to logical and and or, as here. Though they are a bit ambiguous.